File: PHP/test/SomeParsersTest.php
<?php
declare(strict_types=1);

namespace MonadParser\Parser\Test;

include __DIR__ . "/../src/Maybe/Maybe.php";
include __DIR__ . "/../src/Maybe/Constructor/Just.php";
include __DIR__ . "/../src/Maybe/Constructor/Nothing.php";
include __DIR__ . "/../src/Parser.php";
include __DIR__ . "/../src/SomeParsers.php";

use MonadParser\Maybe\Maybe;
use MonadParser\Parser\Parser;
use MonadParser\Parser\SomeParsers;

class SomeParsersTest extends \PHPUnit\Framework\TestCase
{
    public function test_anyChar()
    {
        $this->assertEquals(Parser::anyChar()->parse("abc"), Maybe::just(['a', "bc"]),
            'anyChar.parse("abc") is not equal to Just([\'a\', "bc"])');
        $this->assertEquals(Parser::anyChar()->parse(""), Maybe::nothing(),
            'anyChar.parse("") is not equal to Nothing');
    }

    public function test_satisfy()
    {
        $this->assertEquals(Parser::satisfy(function (string $c){ return $c == 'a'; })->parse("abc"), Maybe::just(['a', "bc"]),
            'satisfy(c==\'a\').parse("abc") is not equal to Just([\'a\', "bc"])');
        $this->assertEquals(Parser::satisfy(function (string $c){ return $c == 'z'; })->parse("abc"), Maybe::nothing(),
            'satisfy(c==\'z\').parse("abc") is not equal to Nothing');
    }

    public function test_char()
    {
        $this->assertEquals(SomeParsers::char_('a')->parse("abc"), Maybe::just(['a', "bc"]),
            'char(\'a\').parse("abc") is not equal to Just([\'a\', "bc"])');
        $this->assertEquals(SomeParsers::char_('z')->parse("abc"), Maybe::nothing(),
            'char(\'z\').parse("abc") is not equal to Nothing');
    }

    public function test_symbol()
    {
        $this->assertEquals(SomeParsers::symbol('+')->parse(" + abc"), Maybe::just(['+', "abc"]),
            'symbol(\'+\').parse(" + abc") is not equal to Just([\'+\', "abc"])');
        $this->assertEquals(SomeParsers::symbol('+')->parse("abc"), Maybe::nothing(),
            'char(\'+\').parse("abc") is not equal to Nothing');
    }

    public function test_empty_string()
    {
      $this->assertEquals(Parser::empty_string()->parse("abc"), Maybe::just(["", "abc"]),
        'empty_string.parse("abc") is not equal to Just(["", "abc"])');
    }

    public function test_optional()
    {
        $this->assertEquals(SomeParsers::optional_s(SomeParsers::char_('1'))->parse("1234"), Maybe::just(["1", "234"]),
            'optional_s(char(\'1\')).parse("1234") is not equal to Just(["1", "234"])');
        $this->assertEquals(SomeParsers::optional_s(SomeParsers::char_('1'))->parse("abc"), Maybe::just(["", "abc"]),
            'optional_s(char(\'1\')).parse("abc") is not equal to Just(["", "abc"])');
    }

    public function test_spaces()
    {
        $spaces = Parser::spaces();
        $this->assertEquals($spaces->parse("abc"), Maybe::just(["", "abc"]),
            'spaces.parse("abc") is not equal to Just(["", "abc"])');
        $this->assertEquals($spaces->parse("  abc"), Maybe::just(["  ", "abc"]),
            'spaces.parse("  abc") is not equal to Just(["  ", "abc"])');
    }

    public function test_alnum()
    {
        $alnum = SomeParsers::alnum();
        $this->assertEquals($alnum->parse("123abc"), Maybe::just(['1', "23abc"]),
            'alnum.parse("123abc") is not equal to Just([\'1\', "23abc"])');
        $this->assertEquals($alnum->parse("_123abc"), Maybe::just(['_', "123abc"]),
            'alnum.parse("_123abc") is not equal to Just([\'_\', "123abc"])');
        $this->assertEquals($alnum->parse("!@#$"), Maybe::nothing(),
            'alnum.parse("!@#$") is not equal to Nothing');

        $this->assertEquals($alnum->some()->parse("123abc"), Maybe::just(["123abc", ""]),
            'alnum.some().parse("123abc") is not equal to Just(["123abc", ""])');
    }

    public function test_name()
    {
        $psin = SomeParsers::name("sin");
      
        $this->assertEquals($psin->parse(" sin "), Maybe::just(["sin", ""]),
            'psin.parse(" sin ") is not equal to Just(["sin", ""])');
        $this->assertEquals($psin->parse(" sin (1.)"), Maybe::just(["sin", "(1.)"]),
            'psin.parse(" sin (1.)") is not equal to Just(["sin", "(1.)"])');
        $this->assertEquals($psin->parse("sinabc"), Maybe::nothing(),
            'psin.parse("sinabc") is not equal to Nothing');
    }

    public function test_digits()
    {
        $digits = SomeParsers::digits();
        
        $this->assertEquals($digits->parse("123abc"), Maybe::just(["123", "abc"]),
            'digits.parse("123abc") is not equal to Just(["123", "abc"])');
        $this->assertEquals($digits->parse("123  abc"), Maybe::just(["123", "  abc"]),
            'digits.parse("123  abc") is not equal to Just(["123", "  abc"])');
        $this->assertEquals($digits->parse("abc"), Maybe::just(["", "abc"]),
            'digits.parse("abc") is not equal to Just(["", "abc"])');
    }

    public function test_sign()
    {
        $sign = SomeParsers::sign();
        $usign = SomeParsers::usign();
        
        $this->assertEquals($sign->parse("abc"), Maybe::just(["", "abc"]),
            'sign.parse("abc") is not equal to Just(["", "abc"])');
        $this->assertEquals($sign->parse("+abc"), Maybe::just(["+", "abc"]),
            'sign.parse("+abc") is not equal to Just(["+", "abc"])');
        $this->assertEquals($sign->parse("-abc"), Maybe::just(["-", "abc"]),
            'sign.parse("-abc") is not equal to Just(["-", "abc"])');

        $this->assertEquals($usign->parse("abc"), Maybe::just(["", "abc"]),
            'usign.parse("abc") is not equal to Just(["", "abc"])');
        $this->assertEquals($usign->parse(" + abc"), Maybe::just(["+", "abc"]),
            'usign.parse("+abc") is not equal to Just(["+", "abc"])');
        $this->assertEquals($usign->parse(" - abc"), Maybe::just(["-", "abc"]),
            'usign.parse("-abc") is not equal to Just(["-", "abc"])');
    }

    public function test_double()
    {
        $double = SomeParsers::double_();
        
        $this->assertEquals($double->parse("1 abc"), Maybe::just([1, "abc"]),
            'double.parse("1 abc") is not equal to Just([1, "abc"])');
        $this->assertEquals($double->parse("1. abc"), Maybe::just([1, "abc"]),
            'double.parse("1. abc") is not equal to Just([1, "abc"])');
        $this->assertEquals($double->parse("1.23 abc"), Maybe::just([1.23, "abc"]),
            'double.parse("1.23 abc") is not equal to Just([1.23, "abc"])');
        $this->assertEquals($double->parse("-1.23 abc"), Maybe::just([-1.23, "abc"]),
            'double.parse("-1.23 abc") is not equal to Just([-1.23, "abc"])');
        $this->assertEquals($double->parse(".23 abc"), Maybe::just([0.23, "abc"]),
            'double.parse(".23 abc") is not equal to Just([0.23, "abc"])');
        $this->assertEquals($double->parse(" + 1.23 abc"), Maybe::nothing(),
            'double.parse(" + 1.23 abc") is not equal to Nothing');
        $this->assertEquals($double->parse("1.23e10abc"), Maybe::just([1.23e10, "abc"]),
            'double.parse("1.23e10abc") is not equal to Just([1.23e10, "abc"])');
        $this->assertEquals($double->parse("1.23e-10abc"), Maybe::just([1.23e-10, "abc"]),
            'double.parse("1.23e-10abc") is not equal to Just([1.23e-10, "abc"])');
        $this->assertEquals($double->parse("abc"), Maybe::nothing(),
            'double.parse("abc") is not equal to Nothing');
    }

    public function test_between()
    {
        $expr = Parser::between(SomeParsers::symbol('('), SomeParsers::symbol(')'),
            function(){ return SomeParsers::double_(); });
      
        $this->assertEquals($expr->parse("( 123 ) abc"), Maybe::just([123, "abc"]),
            'expr.parse(" ( 123 ) abc") is not equal to Just([123, "abc"])');
        $this->assertEquals($expr->parse("( 123 abc"), Maybe::nothing(),
            'expr.parse(" ( 123 abc") is not equal to Nothing');
        $this->assertEquals($expr->parse(" 123 ) abc"), Maybe::nothing(),
            'expr.parse(" 123 ) abc") is not equal to Nothing');
    }

    public function test_chainlr1()
    {
        $add = SomeParsers::symbol('+')->skip(function (){ return Parser::pure(function ($x, $y){ return $x + $y; }); });
        $sub = SomeParsers::symbol('-')->skip(function (){ return Parser::pure(function ($x, $y){ return $x - $y; }); });
        $pexpr = SomeParsers::double_()->chainl1($add->orElse($sub), false);
        $pow = SomeParsers::symbol('^')->skip(function (){ return Parser::pure(function ($x, $y){ return exp($y * log($x)); }); });

        $this->assertEquals($pexpr->parse("7abc"), Maybe::just([7, "abc"]),
            'pexpr.parse("7abc") is not equal to Just([7, "abc"])');
        $this->assertEquals($pexpr->parse("7 - 1 - 2 abc"), Maybe::just([4, "abc"]),
            'pexpr.parse("7 - 1 - 2 abc") is not equal to Just([4, "abc"])');
        $this->assertEquals($pexpr->parse("7 - 1 + 2 - 3 abc"), Maybe::just([5, "abc"]),
            'pexpr.parse("7 - 1 + 2 - 3 abc") is not equal to Just([5, "abc"])');
        $this->assertEquals($pexpr->parse("abc"), Maybe::nothing(),
            'pexpr.parse("abc") is not equal to Nothing');

        $this->assertEquals(SomeParsers::double_()->chainr1($pow)->parse("3 ^ 2 ^ 3 abc")->map(
            function ($pair){ return [round($pair[0]), $pair[1]]; }),
            Maybe::just([6561, "abc"]),
            'double.chainr1(pow).parse("3 ^ 2 ^ 3 abc") is not equal to Just([6561, "abc"])');
    }
}

File: Kotlin/src/test/kotlin/SomeParsersTest.kt
import SomeParsers.Companion.alnum
import SomeParsers.Companion.anyChar
import SomeParsers.Companion.between
import SomeParsers.Companion.chainl1
import SomeParsers.Companion.chainr1
import SomeParsers.Companion.char
import SomeParsers.Companion.digits
import SomeParsers.Companion.double
import SomeParsers.Companion.empty_string
import SomeParsers.Companion.name
import SomeParsers.Companion.optionalC
import SomeParsers.Companion.satisfy
import SomeParsers.Companion.sign
import SomeParsers.Companion.spaces
import SomeParsers.Companion.symbol
import SomeParsers.Companion.usign
import java.util.Optional
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.round
import kotlin.test.Test
import kotlin.test.assertEquals

class SomeParsersTest {
    @Test
    fun `test anyChar` () {
        assertEquals(Optional.of(Pair('a', "bc")), anyChar.parse("abc"))
        assertEquals(Optional.empty(), anyChar.parse(""))
    }

    @Test
    fun `test satisfy` () {
        assertEquals(Optional.of(Pair('a', "bc")), satisfy { c -> c == 'a' }.parse("abc"))
        assertEquals(Optional.empty(), satisfy { c -> c == 'z' }.parse("abc"))
        assertEquals(Optional.empty(), satisfy { c -> c == 'a' }.parse(""))
    }

    @Test
    fun `test char` () {
        assertEquals(Optional.of(Pair('a', "bc")), char('a').parse("abc"))
        assertEquals(Optional.empty(), char('z').parse("abc"))
        assertEquals(Optional.empty(), char('a').parse(""))
    }

    @Test
    fun `test empty_string` () {
        assertEquals(Optional.of(Pair("", "abc")), empty_string.parse("abc"))
    }

    @Test
    fun `test optional` () {
        assertEquals(Optional.of(Pair("1", "234")), optionalC(char('1')).parse("1234"))
        assertEquals(Optional.of(Pair("", "abc")), optionalC(char('1')).parse("abc"))
    }

    @Test
    fun `test spaces` () {
        assertEquals(Optional.of(Pair("", "abc")), spaces.parse("abc"))
        assertEquals(Optional.of(Pair("   ", "abc")), spaces.parse("   abc"))
    }

    @Test
    fun `test symbol` () {
        assertEquals(Optional.of(Pair('+', "abc")), symbol('+').parse(" + abc"))
        assertEquals(Optional.empty(), symbol('+').parse("abc"))
    }

    @Test
    fun `test alnum` () {
        assertEquals(Optional.of(Pair('1', "23abc")), alnum.parse("123abc"))
        assertEquals(Optional.of(Pair('_', "123abc")), alnum.parse("_123abc"))
        assertEquals(Optional.empty(), alnum.parse("!@#"))
    }

    @Test
    fun `test name` () {
        val psin = name("sin")

        assertEquals(Optional.of(Pair("sin", "")), psin.parse(" sin "))
        assertEquals(Optional.of(Pair("sin", "(1.)")), psin.parse(" sin (1.)"))
        assertEquals(Optional.empty(), psin.parse("sinus"))
    }

    @Test
    fun `test sign` () {
        assertEquals(Optional.of(Pair("", "abc")), sign.parse("abc"))
        assertEquals(Optional.of(Pair("+", "abc")), sign.parse("+abc"))
        assertEquals(Optional.of(Pair("-", "abc")), sign.parse("-abc"))

        assertEquals(Optional.of(Pair("", "abc")), usign.parse("abc"))
        assertEquals(Optional.of(Pair("+", "abc")), usign.parse(" + abc"))
        assertEquals(Optional.of(Pair("-", "abc")), usign.parse(" - abc"))
    }

    @Test
    fun `test digits` () {
        assertEquals(Optional.of(Pair("123", "abc")), digits.parse("123abc"))
        assertEquals(Optional.of(Pair("123", "  abc")), digits.parse("123  abc"))
        assertEquals(Optional.of(Pair("", "abc")), digits.parse("abc"))
    }

    @Test
    fun `test double` () {
        assertEquals(Optional.of(Pair(1.0, "abc")), double.parse("1 abc"))
        assertEquals(Optional.of(Pair(1.0, "abc")), double.parse("1. abc"))
        assertEquals(Optional.of(Pair(1.23, "abc")), double.parse("1.23 abc"))
        assertEquals(Optional.of(Pair(0.23, "abc")), double.parse(".23 abc"))
        assertEquals(Optional.empty(), double.parse(" + 1.23 abc"))
        assertEquals(Optional.of(Pair(1.23e10, "abc")), double.parse("1.23e10abc"))
        assertEquals(Optional.of(Pair(1.23e-10, "abc")), double.parse("1.23e-10abc"))
        assertEquals(Optional.empty(), double.parse("abc"))
    }

    @Test
    fun `test between` () {
        val expr = between(symbol('('), symbol(')')) { double }

        assertEquals(Optional.of(Pair(123.0, "abc")), expr.parse(" ( 123 ) abc"))
        assertEquals(Optional.empty(), expr.parse(" ( 123 abc"))
        assertEquals(Optional.empty(), expr.parse(" 123 ) abc"))
    }

    @Test
    fun `test chainlr1` () {
        val add = symbol('+').skip { Parser.pure { x: Double, y: Double -> x + y } }
        val sub = symbol('-').skip { Parser.pure { x: Double, y: Double -> x - y } }
        val pow = symbol('^').skip { Parser.pure { x: Double, y: Double -> exp(y * ln(x)) } }

        val pexpr = chainl1(double, add.or { sub }, false)

        assertEquals(Optional.of(Pair(7.0, "abc")), pexpr.parse("7abc"))
        assertEquals(Optional.of(Pair(4.0, "abc")), pexpr.parse(" 7 - 1 - 2 abc"))
        assertEquals(Optional.of(Pair(5.0, "abc")), pexpr.parse(" 7 - 1 + 2 - 3 abc"))
        assertEquals(Optional.empty(), pexpr.parse("abc"))

        assertEquals(Optional.of(Pair(6561.0, "abc")),
            chainr1(double, pow).parse("3 ^ 2 ^ 3 abc").map { (a, out) -> Pair(round(a), out) })
    }
}
